### PR TITLE
Factor out RPC from Polykey

### LINKED ISSUE #1: RPC message parser can support arbitrary whitespace such as '\n' within and between JSON RPC messages
### Specification

Currently the agnostic RPC message parser separates the messages based on the`}{` boundry., which makes it difficult to read messages since they are in a single line. 

The objective is to improve readability by formatting a message with newlines and whitespaces, and ensuring that the parser is able to handle these characters. 

```ts
// OLD
{some: 'message'}{other: 'message'}...

// new
{some: 'message'}
{other: 'message'}
...
```

This should be possible by adjusting the `separator` parameter which should be able to discard the newline and other whitespace characters when constructing the JsonStreamParser. (Constructing the stream)

### Additional context

Related MatrixAI/Polykey#249 
Related MatrixAI/Polykey#498 

### Tasks

- [ ] 1. Format the message stream to be human-readable by incorporating newlines and whitespace.
- [ ] 2. Modify parser to discard ' ', \n, \t, \r, \f, \v, \u200B, \u00A0.

### LINKED ISSUE #3: RPC Raw Stream Errors (distinguish between transport errors and application errors)
### Specification

We need to review how errors are handled for RPC raw handlers. The transport level error handling needs to be reviewed as well. Error handling between websocket streams and QUIC streams need to be stadardized between each other as well.

There needs to be separation between transport level errors and RPC level ones. I think this is already the case.

### Additional context

* Related: https://github.com/MatrixAI/Polykey/pull/525#discussion_r1272020257
* Related MatrixAI/Polykey#525 
* Related: MatrixAI/js-ws#4 
* https://github.com/MatrixAI/js-ws/issues/1

### Tasks

1. ...
2. ...
3. ...

### LINKED ISSUE #4: Comprehensive integration and concurrent timeout tests for agnostic RPC system
## Description
Investigate and handle the edge cases of concurrent timeouts in the RPC system across WebSocket and QUIC transports. Implement tests to verify the behavior of both client and server when timeouts occur simultaneously, focusing on hard and soft timeout limits.

## Requirements

1. **Concurrent Handling**: Address both client and server-side timeouts occurring at the same time.
2. **Transport Specific**: Conduct tests for WebSocket and QUIC transport layers.
3. **Hard and Soft Limits**: Establish and test hard limits on the server and soft limits on the client.
4. **Infinite Defaults**: Handle and test cases where timeouts are set to infinite by default.

### Tasks

- [ ] **Concurrent Timeouts Test**: Implement tests to simulate and document the behavior when both the client and server time out simultaneously.
- [ ] **Client-Side Timeouts Test**: Add tests and documentation for scenarios where the client times out first.
- [ ] **Server-Side Timeouts Test**: Add tests and documentation for scenarios where the server times out first.
- [ ] **Hard and Soft Timeouts Test**: Create and document tests to validate server-side hard limits and client-side soft limits.
- [ ] **Infinite Timeouts Test**: Implement and document tests for default infinite timeouts on both client and server.

### Acceptance Criteria

- [ ] Successful demonstration of expected behavior when both client and server time out simultaneously.
- [ ] Client-side and server-side timeout scenarios are accurately tested and documented.
- [ ] Hard and soft timeout limits are validated and documented.
- [ ] Default infinite timeouts are tested and can be manually overridden.

### LINKED ISSUE #5: Make `rpc` handlers abstract arrow function properties.
### Specification

Currently the `rpc` handlers are abstract methods that need to be implemented. These should be an abstract arrow function property and handlers should be implemented as arrow functions. This will allow us to use them without rebinding the `this` context.

This should be done during the agent migration PR.

### Additional context

* Should be done during MatrixAI/Polykey#512 
* Related MatrixAI/Polykey#509 .

### Tasks

1. Update the abstract handler's `handle` methods to be abstract arrow function properties.

### LINKED ISSUE #7: Remove timeout grace timer for `RPC` system
### Specification

In the RPC system, specifically the server side. The handler can be cancelled or timed out from the `ctx: ContextTimed`. This will result in a signal that *SHOULD* cause the handler to end immediately. It is possible for the handler to ignore this however.

Currently there is some logic what will wait a certain delay after cancellation and then force close the stream the handler was handling. This logic needs to be removed.

The reasoning is, we should allow for an arbitrary delay in handling a signal. But also, programmatic problems such as deadlocks should not be silently handled by a fail safe such as this. It should be found and dealt with during testing.

### Additional context

- Related MatrixAI/Polykey#552 

### Tasks

1. [ ] Remove grace timeout timer logic from the RPC system.
2. [ ] Remove any tests that test this and update ones that depend on it.

### LINKED ISSUE #10: Customizable Error Filtering for js-rpc through User-Defined toError/fromError Functions
## Description
Enable user-defined error serialization and deserialization through toError and fromError functions in js-rpc. Add optional replacer and reviver parameters to further customize the (de)serialization process, thereby enhancing security by allowing application-specific rules for sensitive data filtering.

## Requirements

1. User-Driven: User should be able to define toError and fromError functions.
2. Optional Parameters: Introduce replacer and reviver for added customization.
3. Application-Specific: Allow injectable rules for data filtering.
4. PK Conformance: Compatibility with existing Polykey workflows.

### Tasks

- [x] Parameterize toError and fromError: Accept user-defined implementations.
- [x] Constructor Implementation: Include toError, fromError, replacer, and reviver in the constructor.
- [x] Unit Testing: Create tests for the new features.
- [x] Documentation: Update js-rpc documentation.

### Acceptance Criteria

- [x] User-defined toError and fromError functions can be successfully implemented.
- [x] Optional replacer and reviver parameters function as expected.
- [x] The system supports application-specific data filtering rules.
- [x] Documentation is updated.

### LINKED ISSUE #11: IdGen function should work per call.
Refactor IdGen to assign IDs on a per-call basis for streamed RPC responses, ensuring that all responses in a stream share the same ID as the originating request.

### Description
In the existing setup, IdGen assigns unique IDs to each individual RPC request. This becomes problematic when streaming multiple responses, as each response should have an ID identical to that of the original request, according to the [JSON RPC Spec](https://www.jsonrpc.org/specification#response_object).

### Tasks

- [ ] 1. Refactor IdGen so that it applies ID on a per-call basis.

### PR DESCRIPTION
**Pull request by CMCDragonkai**
### Description

This PR factors out the RPC functionality out of Polykey. It needs to work with https://github.com/MatrixAI/js-ws. In fact `@matrixai/ws` can be a `devDependency` in order to demonstrate the integration between WS streams and RPC. But this is not set in stone yet. We have to see what is necessary - trial by fire.

TBD - flesh out this spec by writing up all of the relevant classes and modules you're going to need. For example `RPCClient`, `RPCServer`... etc.

### Issues Fixed

- [ ] * Fixes #1 
- [x] * Fixes #3
- [x] * Fixes #4
- [x] * Fixes #5
- [x] * Fixes #7
- [x] * Fixes #10
- [x] * Fixes #11 

### Tasks

Phase 1: Pre-Migration Assessment (Day 1)

1. Review RPC Classes in Polykey:

- [x] Ensure you understand how the existing classes function within Polykey.

2. Identify Dependencies:

- [x]  Determine what dependencies the classes have within Polykey and how to replace or remove them in the standalone library.

3. Assess @matrixai/ws Integration:

- [x] Quickly evaluate whether @matrixai/ws can and should be integrated with the standalone library.

Phase 2: Environment Setup and Initial Migration (Day 1)

1. Branching and Environment Setup:

- [x] Create a new branch for the migration in the standalone library.
- [x] Install necessary dependencies, possibly including @matrixai/ws.

2. Initial Migration:

- [x] Copy the existing RPC classes to the new standalone library.
- [x] Make an initial commit.

Phase 3: Adaptation and Refactoring (Days 2-3)

1. Adapt Classes for Standalone Use:

- [x] Remove or replace any Polykey-specific dependencies or functionalities.

2. Refactor Code:

- [x] Make any necessary code improvements to ensure that the classes function as expected in a standalone environment.

3. Unit Testing:

- [x] Modify existing tests from Polykey for the standalone library or write new ones to ensure everything works as intended.

Phase 4: Integration and Advanced Testing (Day 4)

1. Integration with @matrixai/ws (if applicable):

- [x] Integrate the RPC functionality with @matrixai/ws. - No longer applicable

2. Comprehensive Tests:

- [x] Add concurrent timeout tests and other comprehensive tests.

3. Sanity Check:

- [x] Ensure all components are working as expected in their new environment.

Phase 5: Final Touches and PR 

1. Documentation:

- [x] Update or add inline comments and documentation specific to the standalone library.

2. Linting and Code Style:
- [x] Ensure the code meets the standalone library's linting and style guidelines.

3.  Squash and Rebase Commits:
- [x] Prepare the commit history for a clean merge.

4.  Final Build and Testing:

- [x] Ensure everything is stable.

5. PR and Review:

- [x] Open a PR for the migration and address any comments that come up during the review.

### Final checklist

* [x] Domain specific tests
* [x] Full tests
* [x] Updated inline-comment documentation
* [x] Lint fixed
* [x] Squash and rebased
* [x] Sanity check the final build

### CODE REVIEW

**Comment by CMCDragonkai** [comment]
@addievo does this PR fix all the issues mentioned that this PR fixes?

Did you separate out all the handlers and other abstract classes?

Did you create the excalidraw diagram?

**Comment by CMCDragonkai** [comment]
Also if this is ready for review you must tick of all the tasks mentioned above. And also the final checklist too.

**Comment by CMCDragonkai** [comment]
Also if it fixes all of these:

image

Write a comment listing how each of them were solved, and which part of the code solves them. Also write a comment in each of those issues indicating how they were solved.

**Comment by CMCDragonkai** [comment]
No diagram for js-rpc.

**Comment by addievo** [comment]
Moving on to issues now, overview of issues and complexity

1. Fixes RPC message parser can support arbitrary whitespace such as '\n' within and between JSON RPC messages
Complexity: Moderate
Supporting arbitrary whitespace might require modifying the parser logic to handle different kinds of whitespace characters, like '\n', '\r', '\t', and spaces. JSON itself ignores whitespace, but if your RPC message format has a specific structure, you'll need to make sure that the parser can handle these variations without breaking.

Suggested Approach:
Modify the message parser to strip or appropriately handle different kinds of whitespace. Add unit tests to validate the new behavior.

2. Fixes RPC Raw Stream Errors (distinguish between transport errors and application errors)
Complexity: Moderate to High
Distinguishing between transport and application errors might require some architectural changes. You'll likely need to introduce an error classification system or adapt the existing one.

Suggested Approach:
Create different classes or error codes for transport and application errors and ensure that they are thrown/captured at the appropriate points in your code. Update the error handling routines to distinguish between these errors.

3. Fixes Make rpc handlers abstract arrow function properties
Complexity: Low to Moderate
This issue likely requires refactoring the handler classes to have abstract arrow function properties. This may involve syntactic changes and could affect type definitions.

Suggested Approach:
Refactor the handler classes to make the function properties abstract. Update any associated types or interfaces to reflect the changes.

4. Fixes Comprehensive integration and concurrent timeout tests for agnostic RPC system
Complexity: High
Writing comprehensive integration and concurrency tests is always a time-consuming task. It requires a deep understanding of the system, what it should and shouldn't do, and how it should behave under different conditions.

Suggested Approach:
Write individual tests for each functional part of the system, and then write integration tests that test the system as a whole. For concurrency tests, consider scenarios where multiple operations happen simultaneously and ensure the system behaves as expected.

5. Fixes Remove timeout grace timer for RPC system
Complexity: Moderate
This likely involves removing or modifying existing functionality related to timing out operations within the RPC system. The challenge here is to make sure that removing the grace timer doesn't introduce new issues.

Suggested Approach:
Remove the grace timer code and run comprehensive tests to ensure that the system behaves correctly without it. Make sure to consider edge cases and potential race conditions.

File: tests/rpc/utils/middleware.test.ts [hunk 1]
```diff
@@ -16,19 +18,36 @@ describe('Middleware tests', () => {
     .noShrink();
 
   testProp(
-    'can parse json stream',
+    'can parse json stream and are human readable',
```
Comment (CMCDragonkai) [inline tests/rpc/utils/middleware.test.ts:21]: You should do this as a separate test.

And we want to test all kinds of whitespace, not just newlines. Test for `\t` or spaces and `\v`... etc. Chatgpt all the whitespace relevant characters.

And if you want, use fastcheck to generate those whitespace characters.

**Comment by tegefaulkes** [comment]
Hey, You don't want to be resolving the issues directly. Add a "Fixes #5" line to this PR under issues fixed and it will resolve the issue when merged.

Seccondly, we want to use the conventional commits style of commit messages. For the above commit `resolves issue 5, makes RPC handlers abstract arrow function propeties` you want to format it more like

```
feat: handlers implementations are now abstract arrow functions

Some commentary here if relevant

* Related #5 

[ci skip]
```

Some `feat` could be feat, fix, build, chore, lint. There are no hard and fast rules here though. The top message starts is lowercase unless there is a noun in it, eg. `fix: `RPCServer` something something`. And no full stop.

Check out conventional commits at https://www.conventionalcommits.org/en/v1.0.0/

File: src/handlers/rawHandler.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import type { ContextTimed } from '@matrixai/contexts';
+import type { ReadableStream } from 'stream/web';
+import type { ContainerType, JSONRPCRequest, JSONValue } from '../types';
+import { Handler } from './handler';
+import { ErrorRPCMethodNotImplemented } from '../errors';
+
+export abstract class RawHandler<
```
Comment (CMCDragonkai) [inline src/handlers/rawHandler.ts:7]: This should be `export default RawHandler`.

File: src/handlers/rawHandler.ts
Comment (CMCDragonkai) [inline src/handlers/rawHandler.ts:1]: Should be `RawHandler.ts`.

Comment (CMCDragonkai) [reply src/handlers/rawHandler.ts:1]: Make sure to do it the same across all class exports.

File: src/callers/clientCaller.ts
Comment (CMCDragonkai) [inline src/callers/clientCaller.ts:1]: Should be `ClientCaller.ts`.

File: src/callers/caller.ts
Comment (CMCDragonkai) [inline src/callers/caller.ts:1]: Should be `Caller.ts`.

File: tsconfig.json [hunk 3]
```diff
@@ -1,5 +1,6 @@
 {
   "compilerOptions": {
+    "experimentalDecorators": true,
```
Comment (CMCDragonkai) [inline tsconfig.json:3]: Could be over the `tsconfig.json` from `js-quic`. It already has `experimentalDecorators`.

File: package.json [hunk 4]
```diff
@@ -54,5 +54,14 @@
     "tsconfig-paths": "^3.9.0",
     "typedoc": "^0.23.21",
     "typescript": "^4.9.3"
+  },
+  "dependencies": {
+    "@fast-check/jest": "^1.7.2",
```
Comment (CMCDragonkai) [inline package.json:59]: This should be a `devDependency`.

File: src/errors/errors.ts [hunk 5]
```diff
@@ -0,0 +1,214 @@
+import type { Class } from '@matrixai/errors';
+import type { JSONValue } from '@/types';
```
Comment (CMCDragonkai) [inline src/errors/errors.ts:2]: Use relative imports under `src` directory.

File: package.json [hunk 6]
```diff
@@ -54,5 +54,14 @@
...
+    "@matrixai/async-init": "^1.9.1",
+    "@matrixai/contexts": "^1.2.0",
+    "@matrixai/id": "^3.3.6",
+    "@matrixai/logger": "^3.1.0",
+    "@streamparser/json": "^0.0.17",
+    "ix": "^5.0.0"
```
Comment (CMCDragonkai) [inline package.json:65]: You need `@matrixai/errors` as well since you're using it in the `src`. Install `^1.2.0`.

Comment (CMCDragonkai) [reply package.json:65]: You also need `@matrixai/events` at `^3.0.2`.

File: package.json [hunk 7]
```diff
@@ -54,5 +54,14 @@
     "tsconfig-paths": "^3.9.0",
     "typedoc": "^0.23.21",
     "typescript": "^4.9.3"
+  },
+  "dependencies": {
+    "@fast-check/jest": "^1.7.2",
+    "@matrixai/async-init": "^1.9.1",
+    "@matrixai/contexts": "^1.2.0",
+    "@matrixai/id": "^3.3.6",
```
Comment (CMCDragonkai) [inline package.json:62]: Do you actually use this dependency? And `ix` too?

File: package.json [hunk 8]
```diff
@@ -54,5 +54,14 @@
     "tsconfig-paths": "^3.9.0",
     "typedoc": "^0.23.21",
     "typescript": "^4.9.3"
```
Comment (CMCDragonkai) [inline package.json:56]: Can you reconcile your dependencies with `js-quic` `feature-events` branch?

File: src/utils/middleware.ts [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+    },
+    start: (controller) => {
+      parser.onValue = (value) => {
+        const jsonMessage = messageParser(value.value);
+        const jsonMessageString = JSON.stringify(jsonMessage);
+        const jsonMessageWithNewline = jsonMessageString + '\n';
+        const jsonMessageObject: T = JSON.parse(jsonMessageWithNewline) as T;
+        controller.enqueue(jsonMessageObject);
+        bytesWritten = 0;
+      };
```
Comment (tegefaulkes) [inline src/utils/middleware.ts:49]: I don't understand what's going on here. Why are you adding a `\n` here? Also doing a stringify and parse within the stream here is really expensive.

**Comment by CMCDragonkai** [comment]
@addievo squashing was not done. It needs to be done first before you can tick them off.

image

File: src/utils/middleware.ts (hunk 9 above)
Comment (CMCDragonkai) [reply src/utils/middleware.ts:49]: Was this done to help solve #1? If so, this would not be correct way of solving this problem.

**Comment by CMCDragonkai** [comment]
#4 requires tests on concurrent timeouts between client and server. Just test what happens, it should all end gracefully. If client times out before server, the server exception is just dropped.

**Comment by CMCDragonkai** [comment]
When you solve a task or issue, you must write a comment here or in the OP how you solved it, and what files are relevant.

File: tests/rpc/RPC.test.ts
Comment (amydevs) [inline tests/rpc/RPC.test.ts:1]: tests should be moved from `tests/rpc/` to just `tests/`

File: src/index.ts [hunk 10]
```diff
@@ -0,0 +1,6 @@
+export { default as RPCClient } from './RPCClient';
+export { default as RPCServer } from './RPCServer';
+export * as utils from './utils';
+export * as types from './types';
+export * as errors from './errors/errors';
```
Comment (amydevs) [inline src/index.ts:5]: `export * as errors from './errors';`

File: src/callers/caller.ts [hunk 11]
```diff
@@ -0,0 +1,13 @@
+import type { HandlerType, JSONValue } from '../../src/types';
```
Comment (amydevs) [inline src/callers/caller.ts:1]: `import type { HandlerType, JSONValue } from '../types';`

File: src/errors/sysexits.ts
Comment (amydevs) [inline src/errors/sysexits.ts:1]: @CMCDragonkai not sure, but i think sysexits should only be implemented by Polykey applications and not libraries?

File: src/handlers/handler.ts [hunk 12]
```diff
@@ -0,0 +1,19 @@
+import type { ContainerType, JSONValue } from '../types';
+
+export abstract class Handler<
```
Comment (amydevs) [inline src/handlers/handler.ts:3]: i think we don't use the export on definition keyword
should be:
```ts
abstract class Handler {
    ...
}
export {
    Handler
}
```

File: src/errors/sysexits.ts
Comment (CMCDragonkai) [reply src/errors/sysexits.ts:1]: Yes remove `sysexits`. That's application-specific. Your error tree should look similar to other libraries like `js-quic` or `js-ws`.

File: src/handlers/handler.ts (hunk 12 above)
Comment (CMCDragonkai) [reply src/handlers/handler.ts:3]: This should be at the end `export default Handler;`, and the whole file should be called `Handler.ts`.

Comment (CMCDragonkai) [reply src/handlers/handler.ts:3]: This rule can help enforce it https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/exports-last.md but it's not perfect.

Comment (CMCDragonkai) [reply src/handlers/handler.ts:3]: Actually this https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/prefer-default-export.md! Try it @addievo.

Comment (CMCDragonkai) [reply src/handlers/handler.ts:3]: https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/prefer-default-export.md#any is probably better.

File: package.json (hunk 7 above)
Comment (addievo) [reply package.json:62]: Uh, async-init? or contexts?


Its for the new Create Destroy (or rather just destroy)

File: src/utils/middleware.ts (hunk 9 above)
Comment (addievo) [reply src/utils/middleware.ts:49]: it was

File: src/errors/sysexits.ts
Comment (addievo) [reply src/errors/sysexits.ts:1]: A bunch of the jests and even client side code is based on exit codes, I am unsure how to refactor it properly.
@CMCDragonkai @amydevs

File: src/RPCServer.ts [hunk 13]
```diff
@@ -20,11 +20,11 @@ import { CreateDestroy, ready } from '@matrixai/async-init/dist/CreateDestroy';
...
-import { RawHandler } from '@/handlers/rawHandler';
-import { DuplexHandler } from '@/handlers/duplexHandler';
-import { ServerHandler } from '@/handlers/serverHandler';
-import { ClientHandler } from '@/handlers/clientHandler';
-import { UnaryHandler } from '@/handlers/unaryHandler';
+import { RawHandler } from '@/handlers/RawHandler';
+import { DuplexHandler } from '@/handlers/DuplexHandler';
+import { ServerHandler } from '@/handlers/ServerHandler';
+import { ClientHandler } from '@/handlers/ClientHandler';
+import { UnaryHandler } from '@/handlers/UnaryHandler';
```
Comment (tegefaulkes) [inline src/RPCServer.ts]: FYI, you can only use `@` in the import path in tests. In the source files you must use relative path for imports. It will build, tests will run. But running outside of tests will have runtime errors..

**Comment by addievo** [comment]
ToDo:

- [ ] 1. Implement matrixai/events inrelation to Destroy. 
- [ ] 2. Change errors to not not rely on sysexits, would also have to change jests.
- [ ] 3. Excalidraw Diagram

Previous issues, not necessary for library 

- [ ] 5. Fixing or discarding #1
- [ ] 6. Trying to fix #7

File: src/RPCServer.ts (hunk 13 above)
Comment (addievo) [reply src/RPCServer.ts]: Gotcha, thanks

File: package.json (hunk 7 above)
Comment (CMCDragonkai) [reply package.json:62]: Do you use ix?

File: src/callers/clientCaller.ts
Comment (CMCDragonkai) [reply src/callers/clientCaller.ts:1]: Do you understand that you need to rename the file and also export default?

File: package.json (hunk 7 above)
Comment (addievo) [reply package.json:62]: no

**Comment by CMCDragonkai** [comment]
@amydevs Can you review if the JSON serialisation/encoding makes sense for the current RPC library. Remember that when we use the RPC system in PK or PKE, we would have different application level exceptions that we would throw. Such exceptions would need to be caught and put under the `cause` of the root RPC error before being transmitted to the other side. When the other side receives this chained serialised error, it will try to re-instantiate it as the exception instance, but if it doesn't have one that corresponds it should either leave it as serialised form or instantiate it as an unknown error. I forgot which way we decided before - @tegefaulkes can give some pointers.

Furthermore just before the application error is serialised it should be filtered against rules on the underlying data (including all recursive cause). These rules are supposed to be provided by the application in order to prevent any leakage of sensitive data in the exception chain. This was the case before in GRPC, not sure about RPC - confirm that this is in fact tested an demonstrated that it works.

Finally with everything that has been exported make sure the index.ts is re-exporting everything fractal-like.

File: src/utils/middleware.ts (hunk 9 above)
Comment (CMCDragonkai) [reply src/utils/middleware.ts:49]: Can you answer @tegefaulkes comment above?

File: package.json (hunk 7 above)
Comment (CMCDragonkai) [reply package.json:62]: Then why do you have it?

File: src/errors/sysexits.ts
Comment (addievo) [reply src/errors/sysexits.ts:1]: Resolved by Amy

File: src/utils/middleware.ts (hunk 9 above)
Comment (addievo) [reply src/utils/middleware.ts:49]: > I don't understand what's going on here. Why are you adding a `\n` here? Also doing a stringify and parse within the stream here is really expensive.

Serialises the json to string, adds a new line (wherein it makes it human readable per the bare description of the issue) and then deserialises it back to an object. I now see why it might introduce a lot of overhead, lemme see how else I can resolve it. Maybe before it is sent as a stream would be a good place to change it

**Comment by CMCDragonkai** [comment]
Client service will have no rules and agent service will have strict rules.

File: src/utils/middleware.ts [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+      parser.onEnd = () => waitP.resolveP();
+      parser.end();
+      await waitP.p;
+    },
+    start: (controller) => {
+      parser.onValue = (value) => {
+        const jsonMessage = messageParser(value.value);
+        const jsonMessageString = JSON.stringify(jsonMessage);
+        const jsonMessageWithNewline = jsonMessageString + '\n';
+        const jsonMessageObject: T = JSON.parse(jsonMessageWithNewline) as T;
```
Comment (amydevs) [inline src/utils/middleware.ts:46]: i don't think this is a good idea, JSON.parse(JSON.stringify(...)) is slow

Comment (CMCDragonkai) [reply src/utils/middleware.ts:46]: Already had a conversation with Aditya, this solution is not correct. It should be on the parser side.

File: src/utils/middleware.ts (hunk 9 above)
Comment (tegefaulkes) [reply src/utils/middleware.ts:49]: Ok, I see that the issue is not clear. It does depend on a bunch of context and previous discussion. I'll expand on it in #1 in a moment.

File: src/errors/errors.ts [hunk 15]
```diff
@@ -0,0 +1,215 @@
...
+import { AbstractError } from '@matrixai/errors';
+
+interface RPCError extends Error {}
```
Comment (amydevs) [inline src/errors/errors.ts:5]: remove this line

File: src/errors/errors.ts [hunk 16]
```diff
@@ -0,0 +1,215 @@
...
+
+class ErrorRPC<T> extends Error implements RPCError {
```
Comment (amydevs) [inline src/errors/errors.ts:7]: replace errorCode with just code,
create an enum to reference the standard recommended codes from the spec:
https://www.jsonrpc.org/specification#error_object

File: src/types.ts [hunk 17]
```diff
@@ -0,0 +1,356 @@
...
+   * A Structured value that holds the parameter values to be used during the invocation of the method.
+   *  This member MAY be omitted.
+   */
+  params?: T;
+  /**
+   * An identifier established by the Client that MUST contain a String, Number, or NULL value if included.
+   *  If it is not included it is assumed to be a notification. The value SHOULD normally not be Null [1] and Numbers
+   *  SHOULD NOT contain fractional parts [2]
+   */
+  id: string | number | null;
```
Comment (amydevs) [inline src/types.ts:36]: Lets make id default to `null`

Allow the user to pass in a generator function that returns `string | number | null` in case they want to use a specific id for debugging purposes. We're going to be doing this in PK

File: src/errors/errors.ts (hunk 15 above)
Comment (addievo) [reply src/errors/errors.ts]: resolved

File: src/utils/middleware.ts (hunk 14 above)
Comment (addievo) [reply src/utils/middleware.ts:46]: WIP

File: src/types.ts (hunk 17 above)
Comment (addievo) [reply src/types.ts:36]: wdym?

Comment (CMCDragonkai) [reply src/types.ts:36]: Set the `id` to default to `null`.

Then in `RPCServer` and `RPCClient`, accept a `idGen` function as a parameter that when called gives you a `string | number | null` to put into the ID of every message sent out.

```ts
idGen: () => PromiseLike<string | number | null>
```

Then do `await idGen()`.

Later can be extended to take various input parameters.

File: images/diagram_encapuslated.png
Comment (CMCDragonkai) [inline images/diagram_encapuslated.png:1]: Make your boxes smaller please.

And also when exporting svg, export with background as white, so it is easier to see in git commits.

Comment (CMCDragonkai) [reply images/diagram_encapuslated.png:1]: Export as svg, not png.

Reference the diagram in your `README.md`.

**Review by tegefaulkes — COMMENTED** [review]
`fix` is a keyword in github. If you write `fix #1`. `Fixes #1`, `Fixes: #1` etc etc into a commit message. It will close issue #1 when this commit merges into the main branch. Try to avoid that especially if you're not actually fixing it here.

Same thing is true for Issues and PRs.

Just something to be aware of.

**Comment by CMCDragonkai** [comment]
You can prefer to use `WIP - ...` commit messages first, then rebase squash when ready, and then when we merge, the whole PR system autocloses other issues.

File: src/types.ts [hunk 18]
```diff
@@ -9,6 +9,11 @@ import type { ClientCaller } from './callers';
 import type { UnaryCaller } from './callers';
 import type Handler from './handlers/Handler';
 
+/**
+ * This is the type for the IdGenFunction. It is used to generate the request
+ */
+type IdGenFunction = () => PromiseLike<string | number | null>;
```
Comment (CMCDragonkai) [inline src/types.ts:15]: Rename to just `IdGen`.

File: src/RPCClient.ts [hunk 19]
```diff
@@ -489,11 +489,12 @@ class RPCClient<M extends ClientManifest> {
         abortProm.p,
       ]);
       const tempWriter = rpcStream.writable.getWriter();
+      const id = this.idGen ? await this.idGen() : null;
       const header: JSONRPCRequestMessage = {
         jsonrpc: '2.0',
         method,
         params: headerParams,
-        id: null,
+        id,
```
Comment (CMCDragonkai) [inline src/RPCClient.ts]: Don't do this.

Specify `idGen` as a parameter on creation and constructor.

Give it a default function of `() => null`.

Do the same for `RPCServer`.

File: src/RPCClient.ts [hunk 20]
```diff
@@ -91,7 +91,7 @@ class RPCClient<M extends ClientManifest> {
     logger.info(`Created ${this.name}`);
     return rpcClient;
   }
-
+  protected idGen?: IdGenFunction;
```
Comment (CMCDragonkai) [inline src/RPCClient.ts:94]: This will not be optional. It will always be `protected idGen: IdGen;`. Use defaulted function of `() => null` on the creation.

Don't specify defaults in constructor or in the property initialiser here.

Because it is parameterised, it should only be defaulted in **one** place. That is in the creation under CD and CDSS. Or constructor under SS patterns.

**Comment by CMCDragonkai** [comment]
Notes on point 7. https://github.com/MatrixAI/Polykey/issues/551#issuecomment-1722678331 should be copied into relevant issues @addievo.

Also it wasn't written down there, but timeouts for streams should have 2 timeouts, the default timeout should be the processing of the entire entire stream. And a second optional timeout can be a per-message timeout (per-receved message in particular).

Both should default to `Infinity`. The second timeout is not required under unary calls. In many cases one never bothers to set the second timeout, since connection liveness is already maintained by the underlying transport whether js-ws or js-quic.

Since both timers are built into `RPCClient` and `RPCServer`, then you should have 2 relevant exceptions:

1. `ErrorRPCClientCallerTimeout` - the default timer
2. `ErrorRPCServerHandlerTimeout` - the default timer
3. `ErrorRPCClientCallerIdleTimeout` - an additional option under `ctx`?
4. `ErrorRPCServerHandlerIdleTimeout` - an additional optional parameter under class construction of the handler

Not sure if there's already corresponding exceptions for such things.

Important `data` would be the `method` call.

Another thing is that when timeout occurs, the `reasonToCode` and `codeToReason` should be used to signal to the other side about the stream being timed out. This again would only be done in PK not in the library. But one simply in `js-rpc` throws the error into the respective controllers `controller.error(e)`.

**Comment by CMCDragonkai** [comment]
When @amydevs starts helping here, reserve only a couple hrs for the whitespace problem, it's not entirely critical to the testnet 6 or 7.

File: src/utils/utils.ts [hunk 21]
```diff
@@ -223,32 +223,37 @@ function parseJSONRPCMessage<T extends JSONValue>(
...
+let withinErrorKey = false;
+
+const replacer = (key: string, value: any) => {
+  if (key === 'error') {
+    withinErrorKey = true;
+  }
+  if (withinErrorKey && key !== 'code' && key !== 'message') {
+    if (value instanceof ErrorRPC) {
+      withinErrorKey = false;
+      return {
```
Comment (tegefaulkes) [inline src/utils/utils.ts:236]: The `withinErrorKey` here is essentially a global and will be shared across ALL usages of `replacer`. 

You want to be absolutely certain that's what you want if you're doing this. It's very likely that's undesired and you shouldn't be doing this.

Comment (addievo) [reply src/utils/utils.ts:236]: Good point, I'll handle the state in the function. Thanks!

File: src/utils/utils.ts [hunk 22]
```diff
@@ -0,0 +1,525 @@
...
+  );
+}
+
+/**
+ * Replacer function for serialising errors over RPC (used by `JSON.stringify`
+ * in `fromError`)
+ * Polykey errors are handled by their inbuilt `toJSON` method , so this only
+ * serialises other errors
+ */
+const createReplacer = () => {
```
Comment (amydevs) [inline src/utils/utils.ts:226]: i'm not sure that if the state being encapsulated within an IIFE is the best way to do this, i would imagine it to work better as class

thoughts @CMCDragonkai ?

Comment (CMCDragonkai) [reply src/utils/utils.ts:226]: Not sure but I did notice this utils/utils.ts should be just utils.ts. No need for subdirectory.
